### PR TITLE
Local inference with llama.cpp

### PR DESCRIPTION
**Pull request by zhangir-azerbayev**
This PR adds support for an inference server powered by [llama.cpp](https://github.com/ggerganov/llama.cpp) that runs efficiently on CPU. I am able to get a tactic suggestion in a few seconds on my 5 year old thinkpad. 

See the additions to the README for installation and configuration instructions. 

Unfortunately, the GPTNeoX architecture of the official llmstep model isn't supported by the llama.cpp library. I threw together quick finetune of [open_llama_3b_v2](https://huggingface.co/openlm-research/open_llama_3b_v2), which you can find here: [open-llama-3b_next-tactic_dev0.2](https://huggingface.co/zhangirazerbayev/open-llama-3b_next-tactic_dev0.2). Note this model is extremely undertrained (less than one A100-hour), so should be viewed only as a proof of concept that inference latency on CPU can be acceptable. Training an open-llama that matches the llmstep model will be quite trivial. 

A substantial refactor of `python/train/tune.py` was required train the open llama. But that code is still incredibly messy and I will PR it once I clean it up a bit. 

The reasons this PR is a draft are:
1. The inference server crashes if you move around your cursor too much after typing `llmstep ""`. For now it is best to not move your cursor after typing the empty string, and actually putting text in the string almost always causes a crash. 
2. Currently, this implementation only supports generating one tactic step per `llmstep` call. The CPU won't handle parallel decoding well like the GPU can, but we might want to add a "get another suggestion" button or something of the sort. 
3. We probably want to train the better model compatible with llama.cpp before merging into main.

### CODE REVIEW

**Comment by yangky11** [comment]
Just wondering if the "one tactic suggestion per" is a restriction of llama.cpp? Is it possible to use beam search or other decoding algorithms to get multiple tactics?

**Comment by zhangir-azerbayev** [comment]
> Just wondering if the "one tactic suggestion per" is a restriction of llama.cpp? Is it possible to use beam search or other decoding algorithms to get multiple tactics?

Based on [this section](https://github.com/ggerganov/llama.cpp/blob/master/examples/main/README.md) of the llama.cpp docs, it appears to be a restriction of llama.cpp. In fact, I would suspect it's less a restriction of llama.cpp and more a fundamental limitation of CPU inference, since CPUs probably don't have the FLOPs to do batch level parallelism.

**Comment by wellecks** [comment]
Very exciting! There is a batch size parameter in the docs, could we duplicate the input sequence to get multiple outputs?

**Comment by zhangir-azerbayev** [comment]
I trained a [new tactic generator](https://huggingface.co/zhangirazerbayev/open-llama-3b_next-tactic_dev0.3) for 15 epochs on the llmstep training set. In principle, this model should be competitive with the llmstep model but I haven't extensively checked the training code for bugs that may be impacting performance.

File: README.md [hunk 1]
```diff
@@ -62,6 +62,41 @@ python python/server_vllm.py
...
+# convert the model you downloaded to ggml FP16 format
+python3 convert.py $PATH_TO_MODEL
+```
+This will produce a file named `ggml-model-f32.bin` in the `$PATH_TO_MODEL` directory. 
+
+If the above steps do not work for you, please, consult the more detailed instructions in the [llama.cpp](https://github.com/ggerganov/llama.cpp) repository.
+
+In the following step, you will quantize your model to a reduced precision format. The available formats are `F16, Q8_0, Q5_1, Q5_0, Q4_1, Q4_0`, with lower precision formats trading off accuracy for latency and memory. I would recommend starting with `Q4_0`, and increasing precision if your hardware handles lower precisions comfortably. 
+```bash
+./quantize $PATH_TO_MODEL/ggml-model-f32.bin $PATH_TO_QUANTIZED
```
Comment (kim-em) [inline README.md:92]: This is missing the `type` argument for quantize (e.g. `q4_0`) after the $PATH_TO_QUANTIZED.

File: README.md [hunk 2]
```diff
@@ -62,6 +62,41 @@ python python/server_vllm.py
...
+```
+Then, you may start `server_ggml.py` as follows
+```bash
+python server_ggml.py --model_path $PATH_TO_QUANTIZED
```
Comment (kim-em) [inline README.md]:
```suggestion
python3 python/server_ggml.py --model_path $PATH_TO_QUANTIZED
```

**Comment by kim-em** [comment]
I got this running with the v0.3 model, quantized as q4_0. Unfortunately it is rather unsuccessful on the `Examples.lean`.

**Comment by kim-em** [comment]
I only saw one server crash, a segfault.
